### ISSUE TITLE
DOC: installation of pyICU for macOS

### ISSUE THREAD

**Issue by aqw** [psyinfra/onyo#653]
`onyo get`'s sorting uses `natsort`. And `natsort` on macOS suffers from a maCOS bug with sorting UTF-8.

Overview:
- https://github.com/SethMMorton/natsort/wiki/Possible-Issues-with-natsort.humansorted-or-ns.LOCALE#the-locale-module-is-broken-on-mac-os-x
- https://stackoverflow.com/a/33703122

The solution is to install pyICU.

However, installation has non-python deps.

- https://gist.github.com/ddelange/6e04e81b99fae08e817a00515d4a378d

But it's not clear if that page is isolated to ICU support or something broader.

The goal is to document the installation of ICU and pyICU on macOS. Preferably using only `brew` and/or `pip`.

**Comment by TobiasKadelka** [comment]
To actually install `pyicu` on my macbook, I did the following:

1. 
```
brew uninstall --ignore-dependencies icu4c 
brew install pkg-config icu4c  # keg-only
```

2. 
```
# If you need to have icu4c first in your PATH, run:
echo 'export PATH="/opt/homebrew/opt/icu4c/bin:$PATH"' >> ~/.zshrc
echo 'export PATH="/opt/homebrew/opt/icu4c/sbin:$PATH"' >> ~/.zshrc

# For compilers to find icu4c you may need to set:
export LDFLAGS="-L/opt/homebrew/opt/icu4c/lib"
export CPPFLAGS="-I/opt/homebrew/opt/icu4c/include"

# For pkg-config to find icu4c you may need to set:
export PKG_CONFIG_PATH="/opt/homebrew/opt/icu4c/lib/pkgconfig"

brew install pkg-config icu4c  
```

3.
```
pip uninstall pyicu 

export PATH="/usr/local/opt/icu4c/bin:/usr/local/opt/icu4c/sbin:$PATH"
export PKG_CONFIG_PATH="$PKG_CONFIG_PATH:/usr/local/opt/icu4c/lib/pkgconfig"

# ensure system clang is used for proper libstdc++ https://github.com/ovalhub/pyicu/issues/5#issuecomment-291631507
unset CC CXX

pip install --no-binary=:pyicu: pyicu  # avoid wheels from previous runs or PyPI
```

**Comment by aqw** [comment]
Steps 1 and 2 are two different ways of doing the same thing. So at minimum, it needs to be understood which method is necessary in order to lead to a successful `pip install --no-binary=:pyicu: pyicu`

**Comment by TobiasKadelka** [comment]
This issue is currently blocked by #742. The description for how to install pyICU for macOS given above is also not working anymore for me.

So this issue either needs to wait on how to resolve #742, or it might be closed in favor of it, since it is not clear if there will ever be a good way to install pyICU that can be documented, and we might want to try figure out something else entirely. (I lean into the direction to leave this open until we know more though.)

**Comment by aqw** [comment]
Ok, the process is (modified a bit from https://pypi.org/project/PyICU/):

```
# install libicu (keg-only)
brew install pkg-config icu4c

# let setup.py discover keg-only icu4c via pkg-config
export PATH="/opt/homebrew/opt/icu4c@76/bin:$PATH"
export PATH="/opt/homebrew/opt/icu4c@76/sbin:$PATH"
export PKG_CONFIG_PATH="/opt/homebrew/opt/icu4c/lib/pkgconfig"
```
Then:
```
# Match the compiler used for your Python install
# GCC, since we're using Python from HomeBrew
export CC="$(which gcc)" CXX="$(which g++)"

# install
# avoid wheels from previous runs or PyPI
pip install --no-binary=:pyicu: pyicu
```

If one encounters the following error (which seems to be common):
```
/opt/homebrew/opt/icu4c/include/unicode/unistr.h:35:10: fatal error: 'cstddef' file not found
```
Then chances are that your XCode install is borked. This seems to happen with regularity with macOS upgrades.
```
# install XCode
xcode-select --install
```
Once upon a time, this would guarantee a reinstall. However, it seems that it now says the equivalent of "Ehh; Already installed. Exiting. No override."
Soooo we go nuts and punish XCode for its sins, and remove all of it.
```
# delete the crazy pile of multiple half-broken SDKs that may have accumulated
sudo rm -Rvf /Library/Developer/CommandLineTools/*
```
And now `xcode-select --install` should correctly launch the install GUI. Click through etc. Once completed, the following should work: `pip install --no-binary=:pyicu: pyicu`.

**Comment by aqw** [comment]
Ahh, one last thing: @TobiasKadelka please try running the troublesome `onyo get` in a new terminal.

I want to make sure that `PATH` is completely fresh. It's not clear to me whether `pyicu` is fine on its own, once it's compiled, or if it needs to know the location of `icu4c`. If the latter, then the following will need to be added to `.zshrc` or `.bashrc`:
```
export PATH="/opt/homebrew/opt/icu4c/bin:$PATH"
export PATH="/opt/homebrew/opt/icu4c/sbin:$PATH"
export PKG_CONFIG_PATH="/opt/homebrew/opt/icu4c/lib/pkgconfig"
```

**Comment by TobiasKadelka** [comment]
From a new terminal with a new venv I re-ran the `onyo get` command, the error is there again:

```
(onyo_fresh_and_clean) tkadelka@bnbmac73 in ~/inventory on git:master
❱ onyo get --path groups/[censored]/ł[censored] --keys path
ERROR: [Errno 22] Invalid argument
```

**Comment by aqw** [comment]
Ok, so it sounds like the path is needed. Does it work after running the following?:
```
export PATH="/opt/homebrew/opt/icu4c/bin:$PATH"
export PATH="/opt/homebrew/opt/icu4c/sbin:$PATH"
export PKG_CONFIG_PATH="/opt/homebrew/opt/icu4c/lib/pkgconfig"
```

**Comment by TobiasKadelka** [comment]
No, I try to find the line which fixes it after the group meeting.

**Comment by TobiasKadelka** [comment]
I double checked, and I made a mistake earlier. I can open a new terminal, and if onyo 0.5.0 and pyICU are both installed, it works just fine without exporting anything.

And if I build a new virtual env, I don't have to export anything either to install pyICU into it.

**Comment by aqw** [comment]
Thanks!
